velox/exec/tests/TaskTest.cpp

```diff
@@ -581,4 +597,46 @@ TEST_F(TaskTest, singleThreadedCrossJoin) {
   }
 }
 
+// Check if a downstream task can elegantly close all related upstream tasks
+// once the downstream task completes/aborts.
+TEST_F(TaskTest, checkExchangeSourceClosedAfterAbort) {
+  // Create and start a "remote" task
+  std::string remoteTaskId = "mock://task-0";
+  auto remoteTaskPlan =
+      exec::test::PlanBuilder()
+          .tableScan(ROW({"c0", "c1"}, {INTEGER(), VARCHAR()}))
+          .partitionedOutput({}, 1)
+          .planFragment();
+  auto remoteTask = makeTask(remoteTaskId, remoteTaskPlan, 0);
+  remoteTask->start(remoteTask, 1, 5);
+
+  // Create and start a new task consuming the output of the remote task
+  auto plan = exec::test::PlanBuilder()
+                  .exchange(remoteTaskPlan.planNode->outputType())
+                  .partitionedOutput({}, 1)
+                  .planFragment();
+  plan.executionStrategy = core::ExecutionStrategy::kGrouped;
+  auto task = makeTask("task-1", plan, 0);
+  task->start(task, 1, 5);
+
+  // Create a remote source split and add it the new task
+  MockExchangeSource::resetClosedTasks();
+  task->addSplitWithSequence(
+      "0",
+      exec::Split(
+          std::make_shared<facebook::velox::exec::RemoteConnectorSplit>(
+              remoteTaskId),
+          0),
+      0);
+
+  // Check if the task has closed exchangeSource for the remote task after
+  // aborting the task.
+  auto future = task->requestAbort();
+  future.wait();
+  usleep(100); // Wait for 100ms; close of remote exchange source may not happen
```

Comment:
Would it be possible to use waitForTaskCompletion from velox/exec/tests/utils/QueryAssertions.h here? waitForTaskCompletion may need to be modified to support terminal states other than kFinished though.

Reply:
This is certainly doable. We can add something like a `waitForTaskAbort` similar to `waitForTaskCompletion`. However, it seems like a sleep is still necessary in sense that the destructor of `ExchangeClient` (https://github.com/facebookincubator/velox/blob/aaa93800e4b52ea29705ff2f8213c693f4a92a5a/velox/exec/Exchange.cpp#L239) may not be immediately triggered by `exchangeClients_.clear()` during the task abortion (https://github.com/facebookincubator/velox/blob/dacfce2b12f0bd1ca9a7cabf134afa6abd775c77/velox/exec/Task.cpp#L1200), and we need to wait long enough for the exchangeSource to be actually closed. Let me know if you have different opinion.